File: benchmarks/compare/binarytrees/binarytrees.py
# binarytrees.py — allocation churn (Computer Language Benchmarks Game "binary-trees").
# Bottom-up allocate many short-lived 2-field tree nodes, traverse each to a node
# count, and free them. A node is a (left, right) tuple; a leaf is (None, None).
# Prints exactly one stdout line "RESULT=<int>".
#
# RESULT = stretchCheck + (sum of all iteration checks) + longLivedCheck.
# Parameters (identical across all languages): MIN_DEPTH=4, MAX_DEPTH=16.
import sys

MIN_DEPTH = 4
MAX_DEPTH = 16
sys.setrecursionlimit(100000)


def make(d):
    if d == 0:
        return (None, None)
    return (make(d - 1), make(d - 1))


def check(t):
    l, r = t
    if l is None:
        return 1
    return 1 + check(l) + check(r)


def main():
    max_depth = max(MAX_DEPTH, MIN_DEPTH + 2)
    stretch_check = check(make(max_depth + 1))
    long_lived = make(max_depth)

    total = stretch_check
    for depth in range(MIN_DEPTH, max_depth + 1, 2):
        iterations = 1 << (max_depth - depth + MIN_DEPTH)
        s = 0
        for _ in range(iterations):
            s += check(make(depth))
        total += s

    total += check(long_lived)
    print(f"RESULT={total}")


if __name__ == "__main__":
    main()

File: benchmarks/compare/binarytrees/binarytrees.js
'use strict';
// binarytrees.js — allocation / GC churn (Computer Language Benchmarks Game
// "binary-trees"). Bottom-up allocate many short-lived 2-field objects, traverse
// each to a node count, and let the GC reclaim them. All counts stay within
// Number's safe-integer range (no BigInt). Prints "RESULT=<int>".
//
// RESULT = stretchCheck + (sum of all iteration checks) + longLivedCheck.
// Parameters (identical across all languages): MIN_DEPTH=4, MAX_DEPTH=16.
const MIN_DEPTH = 4;
const MAX_DEPTH = 16;

function make(d) {
  if (d === 0) return { l: null, r: null };
  return { l: make(d - 1), r: make(d - 1) };
}

function check(t) {
  if (t.l === null) return 1;
  return 1 + check(t.l) + check(t.r);
}

function main() {
  const maxDepth = Math.max(MAX_DEPTH, MIN_DEPTH + 2);
  const stretchCheck = check(make(maxDepth + 1));
  const longLived = make(maxDepth);

  let total = stretchCheck;
  for (let depth = MIN_DEPTH; depth <= maxDepth; depth += 2) {
    const iterations = 1 << (maxDepth - depth + MIN_DEPTH);
    let s = 0;
    for (let i = 0; i < iterations; i++) s += check(make(depth));
    total += s;
  }

  total += check(longLived);
  console.log(`RESULT=${total}`);
}

main();
